src/spec-lang/tc/typeenv.ts

```diff
@@ -1,24 +1,98 @@
-import { assert, ContractDefinition, TypeNameType, TypeNode } from "solc-typed-ast";
+import { assert, ContractDefinition, SourceUnit, TypeNameType, TypeNode } from "solc-typed-ast";
 import { ABIEncoderVersion } from "solc-typed-ast/dist/types/abi";
-import { SNode, SUserFunctionDefinition } from "../ast";
+import { SNode, SUserConstantDefinition, SUserFunctionDefinition } from "../ast";
 
 export type TypeMap = Map<SNode, TypeNode>;
 
+/**
+ * Abstract class to map user definitions in Scribble annotations to their scoping areas.
+ */
+export abstract class UserDefinitionScoping<
+    ScopeT extends ContractDefinition | SourceUnit,
+    DefT extends SUserConstantDefinition | SUserFunctionDefinition
+> {
+    readonly mapping: Map<ScopeT, Map<string, DefT>> = new Map();
+
+    abstract get(scope: ScopeT, name: string): DefT | undefined;
+
+    define(scope: ScopeT, def: DefT): void {
+        let defs = this.mapping.get(scope);
+
+        if (defs === undefined) {
+            defs = new Map();
+        }
+
+        defs.set(def.name.name, def);
+
+        this.mapping.set(scope, defs);
+    }
+}
+
+/**
+ * Class to map user-defined constants to their scoping areas
+ */
+export class UserConstantScoping extends UserDefinitionScoping<
+    SourceUnit,
+    SUserConstantDefinition
+> {
+    get(scope: SourceUnit, name: string): SUserConstantDefinition | undefined {
+        const defs = this.mapping.get(scope);
```

Comment:
I wonder if we should allow constants defined in base classes to be used in the child classes? In which case the code for this would be the same as the UserFunctionScoping? Lets discuss offline

Reply:
Fixed in https://github.com/ConsenSys/scribble/pull/180/commits/6a49ac80f8d95f333b842793128fd91161c224a6 to allow constants in base contracts to be used in inheriting contracts. Also changed the behavior so that constants are only per contract such that we're consistent with user functions.